File: haskell/test/Corpus.hs
-- | The corpus reader and the entry check (DOCS.md §4.5, §15).
--
-- THE PLUGIN LIBRARY MUST NEVER BE USED TO IMPLEMENT ITS OWN TESTS
-- (AGENTS.md prime directive 6), and that covers the TYPING as well as
-- the comparison. This file asks "Value" what a value is, because
-- "Value" is the JSON reader rather than the library under test — but
-- @same@, @truthy@ and the merge semantics the corpus pins are all
-- re-derived here rather than borrowed from "Types".
--
-- NO TEST FRAMEWORK: §16 permits one runtime dependency, Haskell has no
-- port of it, and HUnit or hspec would be a second. The runner is a
-- @main@ that counts. THE CORPUS IS A VALUE THREADED THROUGH, not a
-- top-level 'IORef': a global would need @unsafePerformIO@ and a
-- NOINLINE pragma to be even approximately correct, which is a lot of
-- machinery for "read a file once".

module Corpus
  ( Subject, Tally(..)
  , loadCorpus, corpusSection, corpusLabel
  , corpusEqual, corpusMatches, regexLite
  , runSection
  ) where

import Control.Exception (try)
import Control.Monad (forM_, when)
import Data.IORef
import Data.List (isPrefixOf, sort)
import System.Environment (lookupEnv)
import System.Exit (ExitCode (..), exitWith)
import System.IO (hPutStrLn, stderr)
import Types
import Value

type Subject = Value -> IO Value

data Tally = Tally {tEntries :: Int, tFailures :: Int}

die :: String -> IO a
die msg = hPutStrLn stderr msg >> exitWith (ExitFailure 2)

-- | The whole corpus, parsed once. Exits loudly if the JSON is missing
-- or malformed: a runner that reports zero tests as a pass is the
-- failure mode doc/plan/handover.md §4 exists to prevent.
loadCorpus :: IO Value
loadCorpus = do
  env <- lookupEnv "PLUGIN_SPEC"
  let path = case env of Just p | not (null p) -> p; _ -> "../spec/plugin.json"
  text <- readFile path
  v <- either (\e -> die ("haskell: " ++ path ++ " is not valid JSON: " ++ e)) return (parseJson text)
  -- Version 1 turns on strict entry validation in every runner. A
  -- corpus that lost its version marker is a corpus whose shape nobody
  -- checked, so refuse it rather than run against it.
  let version = vget (vget v "PLUGIN") "version"
  when (not (isNum version) || asNum version /= 1) $ die "haskell: unsupported spec version"
  return v

corpusSection :: Value -> String -> IO Value
corpusSection cor name = do
  let s = vget (vget cor "primary") name
  when (not (isMap s)) $ die ("haskell: no such corpus section: " ++ name)
  return s

-- | A stable label, so a failure names the entry rather than an index.
corpusLabel :: String -> Int -> Value -> String
corpusLabel group i entry =
  let idv = vget entry "id"
  in if isStr idv then asStr idv else group ++ "#" ++ show i

-- | Deep equality over spec values: key order never matters, list order
-- always does. Written here, not taken from the library.
corpusEqual :: Value -> Value -> Bool
corpusEqual VNull VNull = True
corpusEqual (VBool a) (VBool b) = a == b
corpusEqual (VNum a) (VNum b) = a == b
corpusEqual (VStr a) (VStr b) = a == b
corpusEqual (VList a) (VList b) = length a == length b && and (zipWith corpusEqual a b)
corpusEqual a@(VMap x) b@(VMap y) =
  length x == length y && all (\k -> vhas b k && corpusEqual (vget a k) (vget b k)) (vkeys a)
corpusEqual _ _ = False

-- | A LITERAL-WITH-ANCHORS MATCHER, NOT A REGEX ENGINE.
--
-- GHC's boot libraries carry no regex engine, and §16 permits no second
-- dependency to supply one. Every pattern the corpus writes is a
-- literal, optionally @^@-anchored, so this unescapes and compares —
-- and ERRORS on any unescaped metacharacter, because the one thing a
-- hand-rolled matcher must never do is quietly report a mismatch it
-- could not evaluate.
--
-- Same shape as @lua\/test\/corpus.lua@'s @regexlite@, deliberately.
regexLite :: String -> String -> Bool
regexLite pattern text = go pattern "" False
  where
    go [] acc anchorStart =
      let lit = reverse acc
      in if anchorStart then lit `isPrefixOf` text else lit `infixOf` text
    go ('\\' : c : rest) acc a = go rest (c : acc) a
    go ('^' : rest) acc _
      | null acc && length rest == length pattern - 1 = go rest acc True
    go ('$' : []) acc a =
      let lit = reverse acc
      in if a then text == lit else lit `suffixOf` text
    go (c : rest) acc a
      | c `elem` "*+?()[]{}|." =
          error ("corpus regex needs a real engine, which this port does not have: " ++ pattern)
      | otherwise = go rest (c : acc) a

    infixOf needle hay = any (needle `isPrefixOf`) (tails' hay)
    suffixOf needle hay = reverse needle `isPrefixOf` reverse hay
    tails' [] = [[]]
    tails' s@(_ : xs) = s : tails' xs

-- | @match@ semantics: @__EXISTS__@, @__UNDEF__@, @__NULL__@,
-- @\/regex\/@, and partial map matching. @present@ distinguishes an
-- absent key from one holding null, which @__UNDEF__@ and @__NULL__@
-- exist to tell apart.
corpusMatches :: Value -> Value -> Bool -> Bool
corpusMatches expect actual present
  | isStr expect =
      case asStr expect of
        "__EXISTS__" -> present && not (isNull actual)
        "__UNDEF__" -> not present
        "__NULL__" -> present && isNull actual
        s
          | length s > 2 && head s == '/' && last s == '/' ->
              isStr actual && regexLite (init (tail s)) (asStr actual)
          | otherwise -> structural
  | otherwise = structural
  where
    structural
      | isList expect =
          isList actual && vlen expect == vlen actual
            && and (zipWith (\e a -> corpusMatches e a True) (vitems expect) (vitems actual))
      | isMap expect =
          isMap actual
            && all (\k -> corpusMatches (vget expect k) (vget actual k) (vhas actual k))
                 (sort (vkeys expect))
      | otherwise = corpusEqual expect actual

-- | Run one entry and report the disagreement, or 'Nothing' when it
-- passes.
--
-- The three combinations the spec format allows are enforced here as
-- well as at build time, because a runner that quietly accepted @err@
-- beside @out@ would let a contradictory entry pass.
corpusCheck :: Value -> Subject -> IO (Maybe String)
corpusCheck entry subject
  | haserr && hasout = return (Just "entry has both err and out")
  | not haserr && not hasout && not hasmatch = return (Just "entry asserts nothing")
  | otherwise = do
      outcome <- try (subject entry)
      case outcome of
        Left e -> return (onRaise (e :: PluginError))
        Right value -> return (onValue value)
  where
    haserr = vhas entry "err"
    hasout = vhas entry "out"
    hasmatch = vhas entry "match"

    onRaise e
      | not haserr = Just ("unexpected raise: " ++ peCode e ++ " " ++ peMessage e)
      -- Errors compare by CODE (§12). Message wording is a port's own
      -- business; pinning it would make every translation a corpus
      -- change.
      | isStr want && peCode e /= asStr want =
          Just ("expected code " ++ asStr want ++ ", got " ++ peCode e ++ " (" ++ peMessage e ++ ")")
      | hasmatch && not (corpusMatches (vget entry "match") got True) =
          Just ("error did not match " ++ renderJson (vget entry "match") ++ ", got " ++ renderJson got)
      | otherwise = Nothing
      where
        want = vget entry "err"
        errv =
          vset (vset (vset (VMap []) "code" (VStr (peCode e))) "message" (VStr (peMessage e)))
            "name" (VStr "PluginError")
        got = vset (VMap []) "err" errv

    onValue value
      | haserr = Just ("expected a raise, got: " ++ renderJson value)
      | hasout && not (corpusEqual (vget entry "out") value) =
          Just ("expected " ++ renderJson (vget entry "out") ++ ", got " ++ renderJson value)
      | hasmatch && not (corpusMatches (vget entry "match") got True) =
          Just ("did not match " ++ renderJson (vget entry "match") ++ ", got out=" ++ renderJson value)
      | otherwise = Nothing
      where
        got = vset (vset (VMap []) "in" (vget entry "in")) "out" value

runSection :: Value -> IORef Tally -> String -> (String -> Maybe Subject) -> IO ()
runSection cor tally sec lookupSubject = do
  groups <- corpusSection cor sec
  -- SORTED, so a failure names the same group in the same place on
  -- every run.
  forM_ (sort (vkeys groups)) $ \name ->
    case lookupSubject name of
      -- A group the runner does not know is a group silently not run,
      -- which is worse than a failure.
      Nothing -> do
        modifyIORef' tally (\t -> t {tFailures = tFailures t + 1})
        putStrLn (sec ++ "/" ++ name ++ ": no subject for this group")
      Just s -> runGroup tally sec name (vget groups name) s

runGroup :: IORef Tally -> String -> String -> Value -> Subject -> IO ()
runGroup tally sec group entries subject = do
  let set = vget entries "set"
  when (isList set) $
    forM_ (zip [0 ..] (vitems set)) $ \(i, entry) -> do
      modifyIORef' tally (\t -> t {tEntries = tEntries t + 1})
      why <- corpusCheck entry subject
      case why of
        Nothing -> return ()
        Just msg -> do
          modifyIORef' tally (\t -> t {tFailures = tFailures t + 1})
          -- The label is the entry's own @id@ when it has one, and
          -- those already carry the section — printing the section
          -- again would read @ref\/ref\/canon#trailing@.
          let l = corpusLabel group i entry
          putStrLn (if vhas entry "id" then l ++ ": " ++ msg else sec ++ "/" ++ l ++ ": " ++ msg)

File: c/test/corpus.c
/* The corpus reader and the entry check. See corpus.h. */

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <regex.h>

#include "corpus.h"

static Value *loaded = NULL;

static const char *specpath(void) {
  const char *env = getenv("PLUGIN_SPEC");
  if (NULL != env && '\0' != env[0]) return env;
  return "../spec/plugin.json";
}

Value *corpus(void) {
  if (NULL != loaded) return loaded;

  const char *path = specpath();
  FILE *f = fopen(path, "rb");
  if (NULL == f) {
    fprintf(stderr, "c: cannot open %s\n", path);
    exit(2);
  }
  fseek(f, 0, SEEK_END);
  long size = ftell(f);
  fseek(f, 0, SEEK_SET);
  if (0 > size) {
    fprintf(stderr, "c: cannot size %s\n", path);
    exit(2);
  }
  char *text = (char *)malloc((size_t)size + 1);
  if (NULL == text) {
    fprintf(stderr, "c: out of memory reading %s\n", path);
    exit(2);
  }
  size_t got = fread(text, 1, (size_t)size, f);
  text[got] = '\0';
  fclose(f);

  const char *err = NULL;
  Value *v = vparse(text, &err);
  free(text);
  if (NULL == v) {
    fprintf(stderr, "c: %s is not valid JSON: %s\n", path, err);
    exit(2);
  }

  /* Version 1 turns on strict entry validation in every runner. A
   * corpus that lost its version marker is a corpus whose shape nobody
   * checked, so refuse it rather than run against it. */
  Value *plugin = vget(v, "PLUGIN");
  Value *version = vget(plugin, "version");
  if (!visnum(version) || 1 != (int)vasnum(version)) {
    fprintf(stderr, "c: unsupported spec version\n");
    exit(2);
  }

  loaded = v;
  return loaded;
}

Value *corpus_section(const char *name) {
  Value *primary = vget(corpus(), "primary");
  Value *section = vget(primary, name);
  if (!vismap(section)) {
    fprintf(stderr, "c: no such corpus section: %s\n", name);
    exit(2);
  }
  return section;
}

const char *corpus_label(const char *group, size_t i, Value *entry) {
  Value *id = vget(entry, "id");
  size_t sz = strlen(group) + 64;
  if (visstr(id)) sz += strlen(vasstr(id));
  char *out = (char *)arena_alloc(sz);
  if (visstr(id)) snprintf(out, sz, "%s", vasstr(id));
  else snprintf(out, sz, "%s#%zu", group, i);
  return out;
}

bool corpus_equal(Value *a, Value *b) {
  bool anull = visnull(a), bnull = visnull(b);
  if (anull || bnull) return anull && bnull;
  if (a->kind != b->kind) return false;
  switch (a->kind) {
    case VBOOL: return vasbool(a) == vasbool(b);
    case VNUM: return vasnum(a) == vasnum(b);
    case VSTR: return 0 == strcmp(vasstr(a), vasstr(b));
    case VLIST: {
      if (vlen(a) != vlen(b)) return false;
      for (size_t i = 0; i < vlen(a); i++) {
        if (!corpus_equal(vat(a, i), vat(b, i))) return false;
      }
      return true;
    }
    case VMAP: {
      if (vlen(a) != vlen(b)) return false;
      const char **keys;
      size_t n = vkeys(a, &keys);
      for (size_t i = 0; i < n; i++) {
        if (!vhas(b, keys[i])) return false;
        if (!corpus_equal(vget(a, keys[i]), vget(b, keys[i]))) return false;
      }
      return true;
    }
    default: return true;
  }
}

bool corpus_matches(Value *expect, Value *actual, bool present) {
  if (visstr(expect)) {
    const char *s = vasstr(expect);
    if (0 == strcmp(s, "__EXISTS__")) return present && !visnull(actual);
    if (0 == strcmp(s, "__UNDEF__")) return !present;
    if (0 == strcmp(s, "__NULL__")) return present && visnull(actual);

    size_t n = strlen(s);
    if (2 < n && '/' == s[0] && '/' == s[n - 1]) {
      if (!visstr(actual)) return false;
      char *pattern = arena_strndup(s + 1, n - 2);
      regex_t re;
      if (0 != regcomp(&re, pattern, REG_EXTENDED)) return false;
      int rc = regexec(&re, vasstr(actual), 0, NULL, 0);
      regfree(&re);
      return 0 == rc;
    }
  }

  if (vislist(expect)) {
    if (!vislist(actual) || vlen(expect) != vlen(actual)) return false;
    for (size_t i = 0; i < vlen(expect); i++) {
      if (!corpus_matches(vat(expect, i), vat(actual, i), true)) return false;
    }
    return true;
  }

  if (vismap(expect)) {
    if (!vismap(actual)) return false;
    const char **keys;
    size_t n = vsortedkeys(expect, &keys);
    for (size_t i = 0; i < n; i++) {
      bool has = vhas(actual, keys[i]);
      if (!corpus_matches(vget(expect, keys[i]), vget(actual, keys[i]), has)) {
        return false;
      }
    }
    return true;
  }

  return corpus_equal(expect, actual);
}

static const char *msgf(const char *fmt, ...) {
  va_list ap;
  va_start(ap, fmt);
  char probe[1];
  va_list ap2;
  va_copy(ap2, ap);
  int n = vsnprintf(probe, 1, fmt, ap2);
  va_end(ap2);
  char *out = (char *)arena_alloc((size_t)n + 1);
  vsnprintf(out, (size_t)n + 1, fmt, ap);
  va_end(ap);
  return out;
}

const char *corpus_check(Value *entry, Subject subject, void *ctx) {
  bool haserr = vhas(entry, "err");
  bool hasout = vhas(entry, "out");
  bool hasmatch = vhas(entry, "match");

  if (haserr && hasout) return "entry has both err and out";
  if (!haserr && !hasout && !hasmatch) return "entry asserts nothing";

  Value *value = NULL;
  PluginError *raised = NULL;

  CatchFrame frame;
  if (0 == PLUGIN_TRY(&frame)) {
    value = subject(entry, ctx);
    PLUGIN_END(&frame);
  }
  else {
    raised = plugin_caught();
  }

  if (haserr) {
    if (NULL == raised) {
      return msgf("expected a raise, got: %s", vjson(value));
    }
    Value *want = vget(entry, "err");
    if (visstr(want)) {
      /* Errors compare by CODE (§12). Message wording is a port's own
       * business; pinning it would make every translation a corpus
       * change. */
      if (0 != strcmp(raised->code, vasstr(want))) {
        return msgf("expected code %s, got %s (%s)",
                    vasstr(want), raised->code, raised->message);
      }
    }
    if (hasmatch) {
      Value *err = vmap();
      vset(err, "code", vstr(raised->code));
      vset(err, "message", vstr(raised->message));
      vset(err, "name", vstr("PluginError"));
      Value *got = vmap();
      vset(got, "err", err);
      if (!corpus_matches(vget(entry, "match"), got, true)) {
        return msgf("error did not match %s, got %s",
                    vjson(vget(entry, "match")), vjson(got));
      }
    }
    return NULL;
  }

  if (NULL != raised) {
    return msgf("unexpected raise: %s %s", raised->code, raised->message);
  }

  if (hasout) {
    if (!corpus_equal(vget(entry, "out"), value)) {
      return msgf("expected %s, got %s",
                  vjson(vget(entry, "out")), vjson(value));
    }
  }

  if (hasmatch) {
    Value *got = vmap();
    vset(got, "in", vget(entry, "in"));
    vset(got, "out", value);
    if (!corpus_matches(vget(entry, "match"), got, true)) {
      return msgf("did not match %s, got out=%s",
                  vjson(vget(entry, "match")), vjson(value));
    }
  }

  return NULL;
}

void corpus_run_group(Tally *t, const char *section, const char *group,
                      Value *entries, Subject subject, void *ctx) {
  Value *set = vget(entries, "set");
  if (!vislist(set)) return;
  for (size_t i = 0; i < vlen(set); i++) {
    Value *entry = vat(set, i);
    t->entries++;
    const char *why = corpus_check(entry, subject, ctx);
    if (NULL != why) {
      t->failures++;
      /* The label is the entry's own `id` when it has one, and those
       * already carry the section — printing the section again would
       * read `ref/ref/canon#trailing`. */
      const char *label = corpus_label(group, i, entry);
      if (vhas(entry, "id")) printf("%s: %s\n", label, why);
      else printf("%s/%s: %s\n", section, label, why);
    }
  }
}

void corpus_run_section(Tally *t, const char *section, Subject subject,
                        void *ctx) {
  Value *groups = corpus_section(section);
  const char **names;
  /* SORTED, so a failure names the same group in the same place on
   * every run. */
  size_t n = vsortedkeys(groups, &names);
  for (size_t i = 0; i < n; i++) {
    corpus_run_group(t, section, names[i], vget(groups, names[i]),
                     subject, ctx);
  }
}
